packages/pool/contracts/StateUtils.sol

```diff
@@ -413,4 +454,70 @@ contract StateUtils is IStateUtils {
             return totalShares();
         }
     }
+
+    /// @notice Called internally to update a checkpoint array
+    /// @param checkpointArray Checkpoint array to be updated
+    /// @param value Value to be updated with
+    function updateCheckpointArray(
+        Checkpoint[] storage checkpointArray,
+        uint256 value
+        )
+        internal
+    {
+        if (checkpointArray.length == 0)
+        {
+            checkpointArray.push(Checkpoint({
+                fromBlock: lastVoteSnapshotBlock,
+                value: value
+                }));
+        }
+        else
+        {
+            Checkpoint storage lastElement = checkpointArray[checkpointArray.length - 1];
+            if (lastElement.fromBlock < lastVoteSnapshotBlock)
+            {
+                checkpointArray.push(Checkpoint({
+                    fromBlock: lastVoteSnapshotBlock,
+                    value: value
+                    }));
+            }
+            else
+            {
+                lastElement.value = value;
+            }
+        }
+    }
+
+    /// @notice Called internally to update an address checkpoint array
+    /// @param addressCheckpointArray Address checkpoint array to be updated
+    /// @param _address Address to be updated with
+    function updateAddressCheckpointArray(
+        AddressCheckpoint[] storage addressCheckpointArray,
+        address _address
+        )
+        internal
+    {
+        if (addressCheckpointArray.length == 0)
+        {
+            addressCheckpointArray.push(AddressCheckpoint({
+                fromBlock: lastVoteSnapshotBlock,
+                _address: _address
+                }));
+        }
+        else
+        {
+            AddressCheckpoint storage lastElement = addressCheckpointArray[addressCheckpointArray.length - 1];
+            if (lastElement.fromBlock < lastVoteSnapshotBlock)
+            {
+                addressCheckpointArray.push(AddressCheckpoint({
+                    fromBlock: lastVoteSnapshotBlock,
+                    _address: _address
+                    }));
+            }
+            else
+            {
+                lastElement._address = _address;
```

Comment:
Does this approach actually works for adressCheckpoint array? We are recording list of delegates, this way if to users delegate in one block we will remember just one of them.

Reply:
A user can have a single delegate at a time, so we're only interested in the most recent delegate. Not sure if you were asking about this.

Reply:
But by this we are removing the latest delegate if he was in this block, aren't we?
By using
`lastElement._address = _address;`

Reply:
Ok, sounds good